src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,318 @@
+using ..Turing
+using ..Bijectors
+using LinearAlgebra
+
+import ..AbstractMCMC: AbstractSampler
+import ..DynamicPPL
+import ..DynamicPPL: Model, AbstractContext, VarInfo, AbstractContext, VarName,
+    _getindex, getsym, getfield, settrans!,  setorder!,
+    get_and_set_val!, istrans, tilde, dot_tilde
+import Optim
+import Optim: optimize
+import NamedArrays
+import ..ForwardDiff
+import StatsBase
+import Printf
+
+export MAP, MLE, optimize
+
+struct MLE end
+struct MAP end
+
+"""
+    OptimizationContext{C<:AbstractContext} <: AbstractContext
+
+The `OptimizationContext` transforms variables to their constrained space, but
+does not use the density with respect to the transformation. This context is
+intended to allow an optimizer to sample in R^n freely.
+"""
+struct OptimizationContext{C<:AbstractContext} <: AbstractContext
+	context::C
+end
+
+# assume
+function DynamicPPL.tilde(ctx::OptimizationContext{<:LikelihoodContext}, spl, dist, vn::VarName, inds, vi)
+    if haskey(vi, vn)
+        # Always overwrite the parameters with new ones for `SampleFromUniform`.
+        if spl isa SampleFromUniform || is_flagged(vi, vn, "del")
+            unset_flag!(vi, vn, "del")
+            r = init(dist, spl)
+            vi[vn] = vectorize(dist, r)
+            settrans!(vi, false, vn)
+            setorder!(vi, vn, get_num_produce(vi))
+        else
+            r = vi[vn]
+        end
+    else
+        r = init(dist, spl)
+        push!(vi, vn, r, dist, spl)
+        settrans!(vi, false, vn)
+    end
+	return r, 0
+end
+
+function DynamicPPL.tilde(ctx::OptimizationContext, spl, dist, vn::VarName, inds, vi)
+    if haskey(vi, vn)
+        # Always overwrite the parameters with new ones for `SampleFromUniform`.
+        if spl isa SampleFromUniform || is_flagged(vi, vn, "del")
+            unset_flag!(vi, vn, "del")
+            r = init(dist, spl)
+            vi[vn] = vectorize(dist, r)
+            settrans!(vi, false, vn)
+            setorder!(vi, vn, get_num_produce(vi))
+        else
+            r = vi[vn]
+        end
+    else
+        r = init(dist, spl)
+        push!(vi, vn, r, dist, spl)
+        settrans!(vi, false, vn)
+    end
```

Comment:
Same here, do we actually need the sampling functionality?
```suggestion
    r = vi[vn]
```